third_party/terraform/utils/retry_transport_test.go

```diff
@@ -0,0 +1,275 @@
+	}
+
+	msg := "body for successful request"
+	ctx, cc := context.WithTimeout(context.Background(), time.Second*2)
+	defer cc()
+
+	req, err := http.NewRequestWithContext(ctx, "GET", ts.URL, bytes.NewReader([]byte(msg)))
+	if err != nil {
+		t.Fatalf("unable to construct err: %v", err)
+	}
+
+	resp, err := client.Do(req)
+	testRetryTransportCheckResponseForSuccess(t, resp, err)
+}
+
+// SUCCESS handlers and check
+func testRetryTransportSucceedAfterHandler(t *testing.T, successInterval time.Duration) http.Handler {
+	var firstReqTime time.Time
+	var testOnce sync.Once
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		testOnce.Do(func() {
+			firstReqTime = time.Now()
+		})
+		if time.Since(firstReqTime) >= successInterval {
+			w.WriteHeader(200)
+			if _, err := w.Write([]byte(testRetryTransportErrorMessageSuccess)); err != nil {
+				t.Errorf("[ERROR] unable to write to response writer: %v", err)
+			}
+		} else {
+			w.WriteHeader(500)
+			if _, err := w.Write([]byte(testRetryTransportErrorMessageRetry)); err != nil {
+				t.Errorf("[ERROR] unable to write to response writer: %v", err)
+			}
+		}
+	})
+}
+
+func testRetryTransportSuccessCheckBodyHandler(t *testing.T, successInterval time.Duration) http.Handler {
+	var firstReqTime time.Time
+	var testOnce sync.Once
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		testOnce.Do(func() {
+			firstReqTime = time.Now()
+		})
+
+		slurp, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			w.WriteHeader(400)
+			if _, err := w.Write([]byte(fmt.Sprintf("unable to read request body: %v", err))); err != nil {
+				t.Errorf("[ERROR] unable to write to response writer: %v", err)
+			}
+			return
+		}
+
+		if time.Since(firstReqTime) >= successInterval {
+			w.WriteHeader(200)
+			resp := fmt.Sprintf("%s\nRequest Body: %s", testRetryTransportErrorMessageSuccess, string(slurp))
+			if _, err := w.Write([]byte(resp)); err != nil {
+				t.Errorf("[ERROR] unable to write to response writer: %v", err)
+			}
+		} else {
+			w.WriteHeader(500)
+			resp := fmt.Sprintf("%s\nRequest Body: %s", testRetryTransportErrorMessageRetry, string(slurp))
+			if _, err := w.Write([]byte(resp)); err != nil {
+				t.Errorf("[ERROR] unable to write to response writer: %v", err)
+			}
+		}
+	})
+}
+
+func testRetryTransportCheckResponseForSuccess(t *testing.T, resp *http.Response, respErr error) {
+	if respErr != nil {
+		t.Fatalf("expected no error, got: %v", respErr)
+	}
+
+	err := googleapi.CheckResponse(resp)
+	if err != nil {
+		t.Fatalf("expected no error, got response error: %v", err)
+	}
+}
+
+// FAILURE handler and check
+func testRetryTransportFailAfterHandler(t *testing.T, successInterval time.Duration) http.Handler {
+	var firstReqTime time.Time
+	var testOnce sync.Once
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		testOnce.Do(func() {
+			firstReqTime = time.Now()
+		})
+		if time.Since(firstReqTime) >= successInterval {
```

Comment:
A FailAfterHandler is basically just a SucceedAfterHandler where the final  error code is 400 instead of 200 and the body is MessageFailure instead of MessageSuccess, right? If so you could make a single AfterHandler func that just takes the error code and message.